### PR TITLE
Reworked bootstrapping notebook

### PR DESCRIPTION
**Pull request by ogrisel**
/cc @glemaitre 

In particular this removes the swarmplot / pd.melt usage and instead uses a scatter plot with blue circle to more intuitively illustrate bootstrap resampling.

It also add bagged polynomial regression pipeline in the end.

### CODE REVIEW

File: python_scripts/ensemble_bagging.py [hunk 1]
```diff
@@ -66,71 +69,62 @@ def generate_data(n_samples=50):
 # %% [markdown]
 # Let's see how we can use bootstraping to learn several trees.
 #
-# ## Bootstrap sample
+# ## Bootstrap resampling
 #
 # A bootstrap sample corresponds to a resampling, with replacement, of the
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:74]: I think we should the first comma. It is difficult to parse the sentence as is.

File: python_scripts/ensemble_bagging.py [hunk 2]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
 #
 # ## Bagging in scikit-learn
 #
-# Scikit-learn implements bagging estimators. It takes a base model that is the
-# model trained on each bootstrap sample.
+# Scikit-learn implements the bagging procedure as a "meta-estimator", that is
+# an estimator that wraps another estimator: it takes a base model that is
+# cloned several times and trained independently on each bootstrap sample.
+#
+# The following shows how to build a bagging ensemble of decision trees. We set
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:224]:
```suggestion
# The following code snippet shows how to build a bagging ensemble of decision trees. We set
```

File: python_scripts/ensemble_bagging.py [hunk 3]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
 
+# %% [markdown]
+#
+# Let us visualize the predictions of the ensemble on the same test data:
 # %%
 sns.scatterplot(x=data_train["Feature"], y=target_train, color="black",
                 alpha=0.5)
-plt.plot(data_test, target_predicted_forest, label="Bag of decision trees")
-plt.legend()
+plt.plot(data_test, bagged_trees.predict(data_test))
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:243]: Just for consistence with the other notebooks where we go a bit slower:

```suggestion
target_predicted = bagged_trees.predict(data_test)
plt.plot(data_test, target_predicted)
```

File: python_scripts/ensemble_bagging.py [hunk 4]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
 
 # %% [markdown]
-# While we used a decision tree as a base model, nothing prevent us of using
-# any other type of model. We will give an example that will use a linear
-# regression.
+#
+# Because we use 100 trees in the ensemble, the average prediction is indeed
+# slightly smoother but very similar to our previous average plot.
+#
+# It is possible to access the internel models of the ensemble in stored as a
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:251]:
```suggestion
# It is possible to access the internal models of the ensemble stored as a
```

File: python_scripts/ensemble_bagging.py [hunk 5]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
-from sklearn.linear_model import LinearRegression
+for tree_idx, tree in enumerate(bagged_trees.estimators_):
+    label = "Predictions of individual trees" if tree_idx == 0 else None
+    plt.plot(data_test, tree.predict(data_test), linestyle="--", alpha=0.1,
+             color="tab:blue", label=label)
+
+sns.scatterplot(x=data_train["Feature"], y=target_train, color="black",
+                alpha=0.5)
+
+plt.plot(data_test, bagged_trees.predict(data_test),
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:265]:
```suggestion
target_predicted = bagged_trees.predict(data_test)
plt.plot(data_test, target_predicted,
```

File: python_scripts/ensemble_bagging.py [hunk 6]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
+# It is possible to access the internel models of the ensemble in stored as a
+# Python list in the `bagged_trees.estimators_` attribute after fitting.
+#
+# Let us compare the based model predictions with their average:
 
 # %%
-from sklearn.linear_model import LinearRegression
+for tree_idx, tree in enumerate(bagged_trees.estimators_):
+    label = "Predictions of individual trees" if tree_idx == 0 else None
+    plt.plot(data_test, tree.predict(data_test), linestyle="--", alpha=0.1,
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:259]:
```suggestion
    target_tree_predicted = tree.predict(data_test)
    plt.plot(data_test, target_tree_predicted, linestyle="--", alpha=0.1,
```

File: python_scripts/ensemble_bagging.py (hunk 1 above)
Comment (ogrisel) [inline python_scripts/ensemble_bagging.py:74]:
```suggestion
# A bootstrap sample corresponds to a resampling with replacement, of the
```

File: python_scripts/ensemble_bagging.py [hunk 7]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
+    MinMaxScaler(),
+    PolynomialFeatures(degree=4),
+    Ridge(alpha=1e-10),
+)
+bagging = BaggingRegressor(
+    base_estimator=polynomial_regressor,
+    n_estimators=100,
+    random_state=0,
+)
+_ = bagging.fit(data_train, target_train)
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:303]:
```suggestion
polynomial_regressor = make_pipeline(
    MinMaxScaler(),
    PolynomialFeatures(degree=4),
    Ridge(alpha=1e-10),
)

# %% [markdown]
# The polynomial regression pipeline will first scale each feature between 0 and 1
# and then expend the feature space by computing polynomial and interaction
# features up to a fourth degree. Now, we can use this model in a bagging ensemble:

# %%
bagging = BaggingRegressor(
    base_estimator=polynomial_regressor,
    n_estimators=100,
    random_state=0,
)
_ = bagging.fit(data_train, target_train)
```

File: python_scripts/ensemble_bagging.py (hunk 3 above)
Comment (ogrisel) [reply python_scripts/ensemble_bagging.py:243]: +1. What I did not like was introducing the local variable in the previous cell and use it only once in a few lines after in the other cell.

File: python_scripts/ensemble_bagging.py (hunk 7 above)
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:303]:
```suggestion
polynomial_regressor = make_pipeline(
    MinMaxScaler(),
    PolynomialFeatures(degree=4),
    Ridge(alpha=1e-10),
)

# %% [markdown]
# The polynomial regression pipeline will first scale each feature between 0 and 1
# and then expend the feature space by computing polynomial and interaction
# features up to a fourth degree. Now, we can use this model in a bagging ensemble:

# %%
bagging = BaggingRegressor(
    base_estimator=polynomial_regressor,
    n_estimators=100,
    random_state=0,
)
_ = bagging.fit(data_train, target_train)
```

File: python_scripts/ensemble_bagging.py [hunk 8]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
+
+
+# %%
+for i, regr in enumerate(bagging.estimators_):
+    base_model_line = plt.plot(
+        data_test, regr.predict(data_test), linestyle="--", alpha=0.2,
+        label="Predictions of base models" if i == 0 else None,
+        color="tab:blue"
+    )
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:315]:
```suggestion
for i, regressor in enumerate(bagging.estimators_):
    target_regressor_predicted = regressor.predict(data_test)
    base_model_line = plt.plot(
        data_test, target_regressor, linestyle="--", alpha=0.2,
        label="Predictions of base models" if i == 0 else None,
        color="tab:blue"
    )
```

File: python_scripts/ensemble_bagging.py [hunk 9]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
+
 sns.scatterplot(x=data_train["Feature"], y=target_train, color="black",
                 alpha=0.5)
-plt.plot(data_test, target_predicted_forest, label="Bag of decision trees")
-plt.plot(data_test, target_predicted_linear, label="Bag of linear regression")
+plt.plot(data_test, bagging.predict(data_test),
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:319]:
```suggestion
target_ensemble_predicted = bagging.predict(data_test)
plt.plot(data_test, target_ensemble_predicted,
```

File: python_scripts/ensemble_bagging.py [hunk 10]
```diff
@@ -184,72 +181,156 @@ def bootstrap_sample(data, target):
...
-_ = plt.title("Bagging classifiers using \ndecision trees and linear models")
+_ = plt.title("Bagged polynomial regression")
 
 # %% [markdown]
-# However, we see that using a bag of linear models is not helpful here because
-# we still obtain a linear model.
+#
+# The predictions of this bagged polynomial regression model looks
+# qualitatively better than the bagged trees. This is somewhat expected since
+# the base models better reflect our knowldge of the true data generating
```
Comment (glemaitre) [inline python_scripts/ensemble_bagging.py:329]:
```suggestion
# the base model better reflects our knowldege of the true data generating
```

File: python_scripts/ensemble_bagging.py (hunk 7 above)
Comment (ogrisel) [reply python_scripts/ensemble_bagging.py:306]: There are no interaction features in this case (single feature dataset).

**Review by glemaitre — APPROVED** [review]
Done with the review.

File: python_scripts/ensemble_bagging.py (hunk 7 above)
Comment (ogrisel) [reply python_scripts/ensemble_bagging.py:306]: I will push an alternative.

**Pull request merged**